services/pgnotifier/pgnotifier.go

```diff
@@ -0,0 +1,447 @@
+	retriggerCount     int
+	trackBatchInterval time.Duration
+)
+
+const (
+	WaitingState   = "waiting"
+	ExecutingState = "executing"
+	SucceededState = "succeeded"
+	FailedState    = "failed"
+	AbortedState   = "aborted"
+)
+
+func init() {
+	queueName = "pg_notifier_queue"
+	maxAttempt = config.GetInt("PgNotifier.maxAttempt", 3)
+	trackBatchInterval = time.Duration(config.GetInt("PgNotifier.trackBatchIntervalInS", 2)) * time.Second
+	retriggerInterval = time.Duration(config.GetInt("PgNotifier.retriggerIntervalInS", 2)) * time.Second
+	retriggerCount = config.GetInt("PgNotifier.retriggerCount", 100)
+}
+
+type PgNotifierT struct {
+	URI      string
+	dbHandle *sql.DB
+}
+
+type MessageT struct {
+	Payload json.RawMessage
+}
+
+type NotificationT struct {
+	ID      int64
+	BatchID string `json:"batch_id"`
+	Status  string
+}
+
+type ResponseT struct {
+	Status  string
+	Payload json.RawMessage
+}
+
+type ClaimT struct {
+	ID                int64
+	BatchID           string
+	Status            string
+	Payload           json.RawMessage
+	ClaimResponseChan chan ClaimResponseT
+}
+
+type ClaimResponseT struct {
+	Payload json.RawMessage
+	Err     error
+}
+
+func New(connectionInfo string) (notifier PgNotifierT, err error) {
+	logger.Infof("PgNotifier: Initializaing PgNotifier...")
+	dbHandle, err := sql.Open("postgres", connectionInfo)
+	if err != nil {
+		return
+	}
+	notifier = PgNotifierT{
+		dbHandle: dbHandle,
+		URI:      connectionInfo,
+	}
+	err = notifier.setupQueue()
+	return
+}
+
+func (notifier PgNotifierT) AddTopic(topic string) (err error) {
+	stmt := fmt.Sprintf("DELETE FROM %s WHERE topic ='%s'", queueName, topic)
+	logger.Infof("PgNotifier: Deleting all jobs on topic: %s", topic)
+	_, err = notifier.dbHandle.Exec(stmt)
+	if err != nil {
+		return
+	}
+	err = notifier.createTrigger(topic)
+	if err != nil {
+		return
+	}
+	rruntime.Go(func() {
+		notifier.triggerPending(topic)
+	})
+	return
+}
+
+func (notifier *PgNotifierT) triggerPending(topic string) {
+	for {
+		time.Sleep(retriggerInterval)
+		stmt := fmt.Sprintf(`UPDATE %[1]s SET status='%[3]s',
+								updated_at = '%[2]s'
+								WHERE id IN  (
+								SELECT id
+								FROM %[1]s
+								WHERE status='%[3]s' OR status='%[4]s'
+								ORDER BY id
+								FOR UPDATE SKIP LOCKED
+								LIMIT %[5]v
+								);`,
+			queueName,
+			GetCurrentSQLTimestamp(),
+			WaitingState,
+			FailedState,
+			retriggerCount)
+		logger.Debugf("PgNotifier: triggering pending jobs")
+		_, err := notifier.dbHandle.Exec(stmt)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (notifier *PgNotifierT) trackBatch(batchID string, ch *chan []ResponseT) {
+	rruntime.Go(func() {
+		for {
+			time.Sleep(trackBatchInterval)
+			// keep polling db for batch status
+			// or subscribe to triggers
+			stmt := fmt.Sprintf(`SELECT count(*) FROM %s WHERE batch_id='%s' AND (status='%s' OR status='%s' OR status='%s')`, queueName, batchID, WaitingState, FailedState, ExecutingState)
```

Comment:
`SELECT count(*) FROM %s WHERE batch_id='%s' AND status != 'Success' AND status !='Aborted')`

This seems better as we just check for terminal states.

Reply:
changed